packages/shared/src/hooks/hook.ts

```diff
@@ -1,17 +1,14 @@
 import { BeforeHookContext, HookContext, HookHints } from './hooks';
-import { EvaluationContext, EvaluationDetails, FlagValue } from '../evaluation';
+import { EvaluationDetails, FlagValue } from '../evaluation';
 
-export interface Hook<T extends FlagValue = FlagValue> {
+export interface BaseHook<T extends FlagValue = FlagValue, BeforeHookReturn = unknown, HooksReturn = unknown> {
```

Comment:
I'm glad you went with a generic solution! Thsi is great!

Interestingly, the client hook generic constraints are a subset of the server hook's constraints. So basically, it seems like you could use the server hook's types here, and then restrict them in the client

```ts
export interface BaseHook<
T extends FlagValue = FlagValue,
BeforeHookReturn = Promise<EvaluationContext | Promise<void>> | EvaluationContext | void,
HooksReturn = Promise<void> | void> {
```

Reply:
Would be okay to me too.
In my view, the current type is more explicit. 
We just have two invariants and both are defined "in-place" where your proposal implies to me that the client is a "special case" and the server is the "normal one".